### ISSUE TITLE
Faster random LongSequence{NucleicAcidAlphabet{4}} generation

### ISSUE THREAD

**Issue by jakobnissen** [BioJulia/BioSequences.jl#77]
I was goofing around the other day and discovered a way to make generation of `LongSequence{NucleicAcidAlphabet{4}}` faster. It's probably fast enough already for any practical application, but in the words of Jeff Bezanson: "[MOAR PERFORMANCE](https://www.youtube.com/watch?v=TPuJsgyu87U&t=338s)". And also, this new code is shorter.
I'm aware that BioSequences guarantees reproducible random sequence generation, and a new algorithm messes with that, so I make this issue here instead of a PR, so we can perhaps use it in a future version of BioSequences.

The new function looks like this:
```
function randseq2(rng::AbstractRNG, A::NucleicAcidAlphabet{4}, len::Integer)
    seq = LongSequence{typeof(A)}(len)
    data = seq.data
    rand!(rng, data)
    @inbounds for i in eachindex(data)
        nuc = 0x1111111111111111
        mask = data[i]
        nuc = ((nuc & mask) << 1) | (nuc & ~mask)
        mask >>>= 1
        nuc = ((nuc & mask) << 2) | (nuc & ~mask)
        data[i] = nuc
    end
    if len > 0
        @inbounds data[end] >>>= ((16 - (len & 15)) & 15) << 2
    end
    return seq
end
```
And it's fast because it uses `rand!` to generate all the randomness at once, and because it SIMD-vectorizes. It's about 8x faster for long sequences.

**Comment by BenJWard** [comment]
What's the loop doing in the middle with the bit flipping?

**Comment by jakobnissen** [comment]
After the call to `rand!`, the sequence data is randomized, which for a `LongSequence{NucleicAcidAlphabet{4}}` means that all IUPAC nucleotides are equally common. The for loops efficiently transforms each 64-bit chunk of random data into an equal distribution of A's, C's, G's and T's instead.
It begins with a sequence of all A's (`0x1111111111111111`). Then, using `mask`, a each bit has a 50% chance of being bitshifted left by one resulting in a 50-50 A/C mix. Then it does another bitshift using the mask, where each A has 50% chance of becoming a G and each C a 50% chance of becoming a T. The mask is shifted to make the first and second shift-rounds independent.

**Comment by BenJWard** [comment]
@jakobnissen Did this get merged?

**Comment by jakobnissen** [comment]
@BenJWard Nope. Looking at the tests for randseq, it appears BioJulia guarantees reproducability (since the actual value of the sequences are tested for), and this change would be breaking. Although I can't find anywhere in the docs that says it's reproducible.

I'd be happy to make a PR implementing this and maybe amending the tests?

**Comment by jakobnissen** [comment]
Actually, since *Julia* doesn't guarantee reproducible RNG streams between minor versions, I'm not sure how BioJulia could guarantee it, unless it rolls its own RNG. Perhaps that restriction should be lifted.

**Comment by jakobnissen** [comment]
After some thought, I think it's better to table this for the time being. Although > 10x faster, breaking the reproducibility is just not worth it. Do you agree, @BenJWard

**Comment by jakobnissen** [comment]
@BenJWard Unfortunately, the 1.5 release makes this issue relevant again. In 1.5, Julia's default RNG has changed its output, meaning BioSequences tests fail. Our choice therefore is to either use StableRNGs.jl as a dependency, or else give up the idea of reproducible biosequences across versions.

Since 1.5 messed up the reproducability anyway, I'll make a PR implementing this.
